core/storage/collection/gae_models.py

```diff
@@ -613,7 +611,7 @@ def convert_to_valid_dict(model_dict: Dict[str, Any]) -> Dict[str, Any]:
     # TODO(#13523): Change 'snapshot_dict' to domain object/TypedDict to
     # remove Any from type-annotation below.
     def _reconstitute(
-            self, snapshot_dict: Dict[str, Any]
+        self, snapshot_dict: Dict[str, Any]
```

Comment:
Can the arg not include Any (e.g. use TypedDict)?

Reply:
While calling this method we are providing JSON values whose type is Any. So that's why Any is used here and there is no domain object which matches this function's argument.

Reply:
Can those JSON values not have type Any? Somewhere down the line there should be a type certainty.

Reply:
Here, we are calling `_reconstitute` method:
https://github.com/oppia/oppia/blob/3c8d9fa916bdd40105a042915162bdccb8ec6a62/core/storage/base_model/gae_models.py#L933

And providing `snapshot_model.content`, which is a JSON property. If we see the description of `content` then it just says JSON blob nothing else.
https://github.com/oppia/oppia/blob/3c8d9fa916bdd40105a042915162bdccb8ec6a62/core/storage/base_model/gae_models.py#L1805

Reply:
> Somewhere down the line there should be a type certainty

From docstring and usage, we are sure that it is going to be a Dict.

Reply:
I see, interesting. Actually, all our JsonProperties have strict types (for a given schema version) so it's a bit of a shame this info gets completely lost. Can we annotate it somehow in the storage layer and assert at that point that it follows the type in question?

That said, I think it's fine to have Any here for now, for the reason you mention. I suggest doing all of the following:

(1) Add a comment here to explain what you just explained above, because that is necessary for justifying why it's not possible to type this strictly.

(2) Discuss with @vojtechjelinek what exactly should be done about this going forward -- ideally we can maintain a type mapping for each JsonProperty in each model that maps the relevant schema version to the expected schema for that version.

(3) File an issue explaining the plan in (2) that both of you come up with, so that we don't lose track of this "hole" in the typing. Add a TODO to that issue here.

Reply:
(1) comment is added, PTAL!
(2) We plan to type each JSON properties explicitly.
(3) TODO is created -- #15911

Reply:
Thanks! I took a look at #15911. Could you please add more detail on how to type each JSON property explicitly (at least roughly)?

Reply:
One example is added in the issue like how we add annotations to JSON properties.

> at least roughly

There is no direct method, which we follow and annotate them -- so that's why an example is added to the issue.